### PR TITLE
Adding code coverage to Cypress (AngularX)

### PR DESCRIPTION
**Pull request by jidma**
## PR description

The PR adds code coverage to Cypress e2e Testing (AngularX front end).

image

When the option is selected, two new npm scripts are added:

The first command allows the user to start the Angular dev server with instrumented code (using istanbul). The dev server is started without `ghostMode`:

    npm run webapp:instrumenter

The second command runs the e2e tests and generates a coverage report:

    npm run e2e:cypress:coverage

Cypress will then run against the dev server on port 9000 (started with npm run start-insrt)

image

After the tests are finished, a report is generated under  `./coverage/lcov-report/`

image

## Known issue

If running against a non instrumented code (ie. started with `npm run e2e:cypress`), Cypress will issue a warning complaining about the code not being instrumented.

## Readme added section:

(...)
### Code Coverage

When using Cypress with AngularX, you can generate code coverage report by running your dev server with instrumented code:

Start your backend (option `-P-webapp` to skip compiling the Angular Front):

    ./mvnw -P-webapp

Start your Angular dev server with instrumented code:

    npm run webapp:instrumenter

Start your Cypress end to end testing:

    npm run e2e:cypress:coverage

The coverage report is generated under `./coverage/lcov-report/`
(...)

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Review by mshima — CHANGES_REQUESTED** [review]
This PR is interesting and you did a great job.
We should make it more modular and simplify it.

File: generators/cypress/index.js [hunk 1]
```diff
@@ -108,4 +109,56 @@ module.exports = class extends BaseBlueprintGenerator {
...
+        message: 'Would you like to generate code coverage for Cypress tests? [Experimental]',
+        default: false,
+      }).then(answers => {
+        this.cypressCoverage = this.jhipsterConfig.cypressCoverage = answers.cypressCoverage;
+        return undefined;
+      });
+    }
+  }
+
+  get prompting() {
```
Comment (mshima) [inline generators/cypress/index.js:134]: Move prompting between initializing and loading (execution order).

**Review by mshima — CHANGES_REQUESTED** [review]

File: generators/cypress/index.js [hunk 2]
```diff
@@ -108,4 +109,56 @@ module.exports = class extends BaseBlueprintGenerator {
     if (useBlueprints) return;
     return this._writing();
   }
+
+  _askForCypressCoverage() {
+    if (this.jhipsterConfig.cypressCoverage !== undefined) {
```
[the remark this answers is not in the document]
Comment (jidma) [reply generators/cypress/index.js:114]: Thank you!

File: generators/client/templates/angular/webpack/webpack.custom.js.ejs [hunk 3]
```diff
@@ -84,6 +84,37 @@ module.exports = (config, options) => {
...
+        }
+      });
+      // We also add istanbul loader
+      config.module.rules.push({
+        test: /\.(js|ts)$/,
+        use: [
+          {
+            loader: 'istanbul-instrumenter-loader',
+            options: { esModules: true },
+          },
```
Comment (mshima) [inline generators/client/templates/angular/webpack/webpack.custom.js.ejs:112]: I just saw that istanbul-instrumenter-loader was archived and conflicts with webpack 5 https://github.com/webpack-contrib/istanbul-instrumenter-loader/issues/110, so no angular 12, react or vue.

This is good for a blueprint, not for inclusion at jhipster.
Can you switch to babel-loader with istanbul plugin?

```suggestion
          {
            loader: 'babel-loader',
            options: {
              plugins: ['istanbul'],
            },
          },
```

File: generators/client/templates/common/package.json [hunk 4]
```diff
@@ -1,7 +1,10 @@
 {
   "devDependencies": {
     "cypress": "6.6.0",
-    "puppeteer": "8.0.0"
+    "puppeteer": "8.0.0",
+    "istanbul-instrumenter-loader": "3.0.1",
```
Comment (mshima) [inline generators/client/templates/common/package.json:5]:
```suggestion
    "babel-loader": "8.2.2",
    "babel-plugin-istanbul": "6.0.0",
```

File: generators/cypress/index.js [hunk 5]
```diff
@@ -108,4 +134,29 @@ module.exports = class extends BaseBlueprintGenerator {
...
     return this._writing();
   }
+
+  _postWriting() {
+    return {
+      configureCoverage() {
+        if (!this.cypressCoverage) return;
+        this.packageJson.merge({
+          devDependencies: {
+            'istanbul-instrumenter-loader': this.configOptions.dependabotPackageJson.devDependencies['istanbul-instrumenter-loader'],
```
Comment (mshima) [inline generators/cypress/index.js:144]:
```suggestion
            'babel-loader': this.configOptions.dependabotPackageJson.devDependencies['babel-loader'],
            'babel-plugin-istanbul': this.configOptions.dependabotPackageJson.devDependencies['babel-plugin-istanbul'],
```

File: generators/common/templates/README.md.ejs [hunk 6]
```diff
@@ -375,6 +375,28 @@ To use those tests, you must install Gatling from [https://gatling.io/](https://
...
+
+    ./mvnw -P-webapp
+
+Start your Angular dev server with instrumented code:
+
+    npm run webapp:instrumenter
+
+Start your Cypress end to end testing:
+
+    npm run e2e:cypress:coverage
```
Comment (mshima) [inline generators/common/templates/README.md.ejs:394]:
```suggestion
    npm run e2e:cypress
```

File: generators/common/templates/README.md.ejs [hunk 7]
```diff
@@ -375,6 +375,28 @@ To use those tests, you must install Gatling from [https://gatling.io/](https://
 
 For more information, refer to the [Running tests page][].
 
+<%_ if (cypressCoverage) { _%>
+
+### Code Coverage
```
Comment (mshima) [inline generators/common/templates/README.md.ejs:380]:
```suggestion
### E2E Webapp Code Coverage
```

File: generators/common/templates/README.md.ejs [hunk 8]
```diff
@@ -375,6 +375,28 @@ To use those tests, you must install Gatling from [https://gatling.io/](https://
...
+
+When using Cypress with AngularX, you can generate code coverage report by running your dev server with instrumented code:
```
Comment (mshima) [inline generators/common/templates/README.md.ejs:382]:
```suggestion
When using Cypress, you can generate code coverage report by running your dev server with instrumented code:
```

File: generators/common/templates/README.md.ejs [hunk 9]
```diff
@@ -375,6 +375,28 @@ To use those tests, you must install Gatling from [https://gatling.io/](https://
...
+
+Start your backend (option `-P-webapp` to skip compiling the Angular Front):
```
Comment (mshima) [inline generators/common/templates/README.md.ejs:384]:
```suggestion
Start your backend without compiling frontend:
```

File: generators/common/templates/README.md.ejs [hunk 10]
```diff
@@ -375,6 +375,28 @@ To use those tests, you must install Gatling from [https://gatling.io/](https://
...
+
+    ./mvnw -P-webapp
```
Comment (mshima) [inline generators/common/templates/README.md.ejs:386]:
```suggestion
    npm run backend:start
```

**Review by mshima — CHANGES_REQUESTED** [review]

File: generators/cypress/index.js [hunk 11]
```diff
@@ -108,4 +134,32 @@ module.exports = class extends BaseBlueprintGenerator {
...
+            '@cypress/code-coverage': this.configOptions.dependabotPackageJson.devDependencies['@cypress/code-coverage'],
+          },
+          scripts: {
+            'prewebapp:instrumenter': 'rimraf .nyc_output && rimraf coverage',
+            'webapp:instrumenter': 'ng build --configuration instrumenter',
+          },
+        });
+        if (this.clientFrameworkAngular) {
+          // Add 'ng build --configuration instrumenter' support
+          this.createStorage('angular.json').merge({projects: {[this_.kebabCase(this.baseName)]: {architect: {build:{configurations:{instrumenter:{}}}}}]}});
```
Comment (mshima) [inline generators/cypress/index.js:155]:
```suggestion
          this.createStorage('angular.json').setPath(`projects.${this._.kebabCase(this.baseName)}.architect.build.configurations.instrumenter`, {});
```

**Comment by jidma** [comment]
@mshima thank once again you for your code review.

After adding and testing your commit, I can't seem to get the instrumented code:

    npm run webapp:instrumenter

    ./mvnw -P-webapp   (because I'm currently on windows and npm run backend:start doesn't work for some reason)

The code doesn't seem to be instrumented:

image

I wanted to let you know first (am I missing something obvious?) before I investigate further later.

**Review by mshima — CHANGES_REQUESTED** [review]
The loader is inside `if (targetOptions.target === 'serve' || config.watch)` but target is `build`.

Rebase against main for webpack needle support.

File: generators/cypress/index.js [hunk 12]
```diff
@@ -48,6 +50,31 @@ module.exports = class extends BaseBlueprintGenerator {
...
+            name: 'cypressCoverage',
+            message: 'Would you like to generate code coverage for Cypress tests? [Experimental]',
+            default: false,
+          }).then(answers => {
+            this.cypressCoverage = this.jhipsterConfig.cypressCoverage = answers.cypressCoverage;
+            return undefined;
+          });
+        }
+      },
+    };
```
Comment (mshima) [inline generators/cypress/index.js:75]:
```suggestion
      async askForCypressCoverage() {
        if (
          this.options.existingProject ||
          this.jhipsterConfig.clientFramework !== constants.SUPPORTED_CLIENT_FRAMEWORKS.ANGULAR ||
          !this.jhipsterConfig.testFrameworks.includes(CYPRESS)
        ) {
          return;
        }
        const answers = await this.prompt({
          type: 'confirm',
          name: 'cypressCoverage',
          message: 'Would you like to generate code coverage for Cypress tests? [Experimental]',
          default: this.jhipsterConfig.cypressCoverage || false,
        });

        this.cypressCoverage = this.jhipsterConfig.cypressCoverage = answers.cypressCoverage;
      },
    };
```
Little cleanup.

**Comment by jidma** [comment]
@mshima it's working now. Thanks!

File: generators/generator-base.js [hunk 13]
```diff
@@ -2372,6 +2372,9 @@ templates: ${JSON.stringify(existingTemplates, null, 2)}`;
     if (options.testFrameworks) {
       this.jhipsterConfig.testFrameworks = options.testFrameworks;
     }
+    if (options.cypressCoverage) {
+      this.jhipsterConfig.cypressCoverage = options.cypressCoverage;
+    }
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/generator-base.js]: Readd this and add the option to the app constructor. To be easier to enable.

File: generators/client/templates/common/package.json [hunk 14]
```diff
@@ -1,7 +1,10 @@
 {
   "devDependencies": {
     "cypress": "6.6.0",
-    "puppeteer": "8.0.0"
+    "puppeteer": "8.0.0",
+    "babel-loader": "8.2.2",
+    "babel-plugin-istanbul": "6.0.0",
+    "@cypress/code-coverage": "3.9.2"
```
Comment (mshima) [inline generators/client/templates/common/package.json:7]:
```suggestion
    "@cypress/code-coverage": "3.9.2",
    "babel-loader": "8.2.2",
    "babel-plugin-istanbul": "6.0.0",
    "cypress": "6.6.0",
    "nyc": "15.1.0",
    "puppeteer": "8.0.0"
```
Reorder and add nyc to print the report.

File: generators/client/templates/common/package.json [hunk 15]
```diff
@@ -1,7 +1,10 @@
 {
   "devDependencies": {
     "cypress": "6.6.0",
```
Comment (mshima) [inline generators/client/templates/common/package.json:3]:
```suggestion
```
Reorder

File: generators/cypress/index.js [hunk 16]
```diff
@@ -108,4 +137,57 @@ module.exports = class extends BaseBlueprintGenerator {
...
+        if (!this.cypressCoverage) return;
+        this.packageJson.merge({
+          devDependencies: {
+            'babel-loader': this.configOptions.dependabotPackageJson.devDependencies['babel-loader'],
+            'babel-plugin-istanbul': this.configOptions.dependabotPackageJson.devDependencies['babel-plugin-istanbul'],
+            '@cypress/code-coverage': this.configOptions.dependabotPackageJson.devDependencies['@cypress/code-coverage'],
+          },
+          scripts: {
+            'prewebapp:instrumenter': 'npm run clean-www && rimraf .nyc_output coverage',
+            'webapp:instrumenter': 'ng build --configuration instrumenter',
```
Comment (mshima) [inline generators/cypress/index.js:153]:
```suggestion
            'clean-coverage': 'rimraf .nyc_output coverage',
            'pree2e:cypress:coverage': 'npm run clean coverage && npm run ci:server:await',
            'e2e:cypress:coverage': 'npm run e2e:cypress',
            'poste2e:cypress:coverage': 'nyc report',
            'prewebapp:instrumenter': 'npm run clean-www && npm run clean-coverage',
            'webapp:instrumenter': 'ng build --configuration instrumenter',
```
Add script to print report.

File: generators/cypress/index.js [hunk 17]
```diff
@@ -108,4 +137,57 @@ module.exports = class extends BaseBlueprintGenerator {
...
+
+  _postWriting() {
+    return {
+      configureCoverage() {
+        if (!this.cypressCoverage) return;
+        this.packageJson.merge({
+          devDependencies: {
+            'babel-loader': this.configOptions.dependabotPackageJson.devDependencies['babel-loader'],
+            'babel-plugin-istanbul': this.configOptions.dependabotPackageJson.devDependencies['babel-plugin-istanbul'],
+            '@cypress/code-coverage': this.configOptions.dependabotPackageJson.devDependencies['@cypress/code-coverage'],
```
Comment (mshima) [inline generators/cypress/index.js:149]:
```suggestion
            '@cypress/code-coverage': this.configOptions.dependabotPackageJson.devDependencies['@cypress/code-coverage'],
            'babel-loader': this.configOptions.dependabotPackageJson.devDependencies['babel-loader'],
            'babel-plugin-istanbul': this.configOptions.dependabotPackageJson.devDependencies['babel-plugin-istanbul'],
            'nyc': this.configOptions.dependabotPackageJson.devDependencies['nyc'],
```

**Review by mshima — APPROVED** [review]
LGTM.

@jidma can you add an option to generator/app/index.js please?
See https://github.com/jhipster/generator-jhipster/pull/14465#discussion_r605132750.

File: generators/generator-base.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (jidma) [reply generators/generator-base.js:2386]: I added the code back but no idea how to add it to the constructor. Will check further later tonight.

[the remark this answers is not in the document]
Comment (mshima) [reply generators/generator-base.js:2386]: Like here https://github.com/jhipster/generator-jhipster/blob/d72f9852bb691b5ae505f8cbdb0595b08d6d990a/generators/app/index.js#L241

**Comment by mshima** [comment]
@jidma congratulations on your first contribution.

**Comment by jidma** [comment]
> @jidma congratulations on your first contribution.

thank you the support!

**Comment by marcolafranca** [comment]
Hi,
thank you for your work.
I have a problem when I enable this generator in jhipster wizard.
Jhipster 7.5.0

image

`ERROR! Error parsing file src\test\javascript\cypress\plugins\index.ts: SyntaxError: ';' expected. (27:4)
  25 |       return launchOptions;
  26 |     }
> 27 |   });
     |    ^
  28 |
  29 |   on('task', {
  30 |     lighthouse: lighthouse((lighthouseReport) => {

At: 1: // ***********************************************************
2: // This example plugins/index.js can be used to load plugins
3: //
4: // You can change the location of this file or turn off loading
5: // the plugins file with the 'pluginsFile' configuration option.
6: //
7: // You can read more here:
8: // https://on.cypress.io/plugins-guide
9: // ***********************************************************
10:
11: // This function is called when a project is opened or re-opened (e.g. due to
12: // the project's config changing)
13: import fs = require('fs');
14: import { lighthouse, pa11y, prepareAudit } from 'cypress-audit';
15: import ReportGenerator = require('lighthouse/report/generator/report-generator');
16: /**
17:  * @type {Cypress.PluginConfig}
18:  */
19: module.exports = (on, config) => {
20:   // `on` is used to hook into various events Cypress emits
21: // `config` is the resolved Cypress config  on('before:browser:launch', (browser, launchOptions) => {
22:     prepareAudit(launchOptions);
23:     if (browser.name === 'chrome' && browser.isHeadless) {
24:       launchOptions.args.push('--disable-gpu');
25:       return launchOptions;
26:     }
27:   });
28:
29:   on('task', {
30:     lighthouse: lighthouse((lighthouseReport) => {
31:       !fs.existsSync('target/cypress') && fs.mkdirSync('target/cypress', { recursive: true });
32:       fs.writeFileSync('target/cypress/lhreport.html', ReportGenerator.generateReport(lighthouseReport.lhr, 'html'));
33:     }),
34:     pa11y: pa11y(),
35:   });
36:     require('@cypress/code-coverage/task')(on, config);
37:     return config;
38: }
39:
Error: Error parsing file src\test\javascript\cypress\plugins\index.ts: SyntaxError: ';' expected. (27:4)
  25 |       return launchOptions;
  26 |     }
> 27 |   });
     |    ^
  28 |
  29 |   on('task', {
  30 |     lighthouse: lighthouse((lighthouseReport) => {

At: 1: // ***********************************************************
2: // This example plugins/index.js can be used to load plugins
3: //
4: // You can change the location of this file or turn off loading
5: // the plugins file with the 'pluginsFile' configuration option.
6: //
7: // You can read more here:
8: // https://on.cypress.io/plugins-guide
9: // ***********************************************************
10:
11: // This function is called when a project is opened or re-opened (e.g. due to
12: // the project's config changing)
13: import fs = require('fs');
14: import { lighthouse, pa11y, prepareAudit } from 'cypress-audit';
15: import ReportGenerator = require('lighthouse/report/generator/report-generator');
16: /**
17:  * @type {Cypress.PluginConfig}
18:  */
19: module.exports = (on, config) => {
20:   // `on` is used to hook into various events Cypress emits
21: // `config` is the resolved Cypress config  on('before:browser:launch', (browser, launchOptions) => {
22:     prepareAudit(launchOptions);
23:     if (browser.name === 'chrome' && browser.isHeadless) {
24:       launchOptions.args.push('--disable-gpu');
25:       return launchOptions;
26:     }
27:   });
28:
29:   on('task', {
30:     lighthouse: lighthouse((lighthouseReport) => {
31:       !fs.existsSync('target/cypress') && fs.mkdirSync('target/cypress', { recursive: true });
32:       fs.writeFileSync('target/cypress/lhreport.html', ReportGenerator.generateReport(lighthouseReport.lhr, 'html'));
33:     }),
34:     pa11y: pa11y(),
35:   });
36:     require('@cypress/code-coverage/task')(on, config);
37:     return config;
38: }
39:
    at patternSpy.dot (C:\Users\marco_ye\AppData\Roaming\npm\node_modules\generator-jhipster\generators\generator-transforms.js:79:15)
    at async PTransform.transform [as _transform] (C:\Users\marco_ye\AppData\Roaming\npm\node_modules\generator-jhipster\node_modules\p-transform\index.js:127:7)
    at async PTransform.queuedTransform (C:\Users\marco_ye\AppData\Roaming\npm\node_modules\generator-jhipster\node_modules\p-transform\index.js:88:26)
    at async run (C:\Users\marco_ye\AppData\Roaming\npm\node_modules\generator-jhipster\node_modules\p-queue\dist\index.js:163:29)
   create src\main\webapp\i18n\en\logs.json
   create src\test\javascript\cypress\fixtures\integration-test.png`

Can you help me? Thank you very much :-)

**Pull request merged**